### ISSUE TITLE
Cannot load TIF image data using E4.25 on Linux

### ISSUE THREAD

**Issue by pagew2000** [eclipse-platform/eclipse.platform.swt#558]
Recently upgraded to Eclipse 4.25. With same tiff image file, it no longer works on Linux.

Here is the code snippet:

    private ImageData getTiffImageData()
    {       
        InputStream stream = this.getClass().getResourceAsStream("testimage.tif");
        ImageData imageData = new ImageData(stream);
        return imageData;
    }

The line new ImageData(stream) throws following exception:

org.eclipse.swt.SWTException: Invalid image  at 
org.eclipse.swt.SWT.error(SWT.java:4918)  at 
org.eclipse.swt.SWT.error(SWT.java:4833)  at 
org.eclipse.swt.SWT.error(SWT.java:4804)  at 
org.eclipse.swt.graphics.ImageLoader.getImageDataArrayFromStream(ImageLoader.java:212)  at 
org.eclipse.swt.graphics.ImageLoader.load(ImageLoader.java:165)  at 
org.eclipse.swt.graphics.ImageDataLoader.load(ImageDataLoader.java:25)  at 
org.eclipse.swt.graphics.ImageData.(ImageData.java:334)

This code works fine when using Eclipse4.16.

**Environment:**
1. Select the platform(s) on which the behavior is seen:
- - [x] Linux

2. JRE/JDK version: JRE11
 
**Version since**
 Eclipse or SWT version since when the behavior is seen: 4.25

**Comment by akurtakov** [comment]
Please attach the image at question so when someone decides to look into details can reproduce it.

**Comment by pagew2000** [comment]
[testimage.zip](https://github.com/eclipse-platform/eclipse.platform.swt/files/10806189/testimage.zip)

**Comment by akurtakov** [comment]
For the record - gnome image viewer and evince both failed to open the tif file in question, gimp succeeded.

**Comment by SyntevoAlex** [comment]
I'll have a look.

**Comment by SyntevoAlex** [comment]
The image is too large and `tiff_image_parse()` in GTK fails to handle it.

Specifically, this code rejects image
```
if (height > G_MAXINT / rowstride) { /* overflow */
	g_set_error_literal (error,
			GDK_PIXBUF_ERROR,
			GDK_PIXBUF_ERROR_CORRUPT_IMAGE,
			_("Dimensions of TIFF image too large"));
	return NULL;
}
```

Where, for your image:
```
width       = 28086
height      = 19866
rowstride   = 112344

width * height * 4 = 0x8506F5F0, more than max 0x7FFFFFFF
```

**Comment by SyntevoAlex** [comment]
Therefore, not a bug in SWT, and not even a bug in GTK. It's just that image is too large.

**Comment by SyntevoAlex** [comment]
I suggest to close the Issue.

**Comment by SyntevoAlex** [comment]
If you can't make the image smaller, then maybe you could use `TIFFFileFormat` from SWT to load image using SWT code. I think that this is what happened when it worked, and then SWT switched to use GTK-provided loader.

**Comment by pagew2000** [comment]
How do you define an image is too big? The sample image size is only at 62kb. If it can only handle an image up to certain size/dimension, can you provide the exact limitations? Why does it work on other platforms?

**Comment by SyntevoAlex** [comment]
It's not "me" but GTK, the Linux UI library, defines it.

I shown the code above. It will fail if `(width * height > 536'870'912)`. In your case, it's `557'956'476`.

**Comment by laeubi** [comment]
@SyntevoAlex what do you think about

```
if (height > G_MAXINT / rowstride) { /* overflow */
  //do not fail but use `TIFFFileFormat`?
}
```

**Comment by SyntevoAlex** [comment]
I have no objections but I'm also lazy to implement this in SWT.

On a side note, anyone interested could try to submit a suggestion to GTK to remove this limitation in 64-bit code. Most likely it was put there for 32-bit code.

**Comment by laeubi** [comment]
> I have no objections but I'm also lazy to implement this in SWT.
> 
> On a side note, anyone interested could try to submit a suggestion to GTK to remove this limitation in 64-bit code. Most likely it was put there for 32-bit code.

I tried on my Debian system with gnome and the default imageviewer fails with exactly the same message "TIF image is too large), but Gimp can open it, ...

**Comment by SyntevoAlex** [comment]
I would say that such a large (geometrically) image shouldn't be kept in TIFF at all. Some kind of vector format, such as SVG, is a much better fit. Otherwise keeping it in memory and painting is too expensive.

**Comment by SyntevoAlex** [comment]
On the other hand, I think that SWT doesn't provide efficient vector painting anyway, it's simply out of scope of SWT's goals.

**Comment by laeubi** [comment]
Yes I once started a bit an "SVG initiative" but its quite complex and has had less interest, beside that such image must not necessarily be more efficient (in terms of speed)or space it all depends on the context.

Nerveless I marked it as regression as SWT previously could handle such images, if its the best choice it nothing we should decide but the application designer.

**Comment by akurtakov** [comment]
SVG can be loaded and rendered on Linux  but is not implemented on other OSes.

**Comment by pagew2000** [comment]
Just FYI, in manufacturing industry, TIFF image is quite common and this size is not considered as large at all. It is quite a regression if we cannot display TIFF on Linux.

**Comment by laeubi** [comment]
> SVG can be loaded and rendered on Linux but is not implemented on other OSes

Yeah but last time i tried this out it looked quite ugly, nothing comparable to a modern SVG render (e.g. browsers)....

**Comment by laeubi** [comment]
@pagew2000 do you like to propose a patch? I think the idea is quite clear here ...

**Comment by pagew2000** [comment]
@laeubi Yes, a patch will be great.

**Comment by mbooth101** [comment]
> > I have no objections but I'm also lazy to implement this in SWT.
> > On a side note, anyone interested could try to submit a suggestion to GTK to remove this limitation in 64-bit code. Most likely it was put there for 32-bit code.
> 
> I tried on my Debian system with gnome and the default imageviewer fails with exactly the same message "TIF image is too large), but Gimp can open it, ...

It is because Gimp has bespoke image loading code and does not use gdk-pixbuf for loading TIFs.
